### PR TITLE
Metamask related fixes

### PR DESCRIPTION
**Pull request by ladislavdubravsky**
Both buy now and auction offerings should now work with both HD wallet and Metamask, for TLD and non-TLD names.

We replaced the workflow of transferring ownership into Offering contract in two transactions by doing it in one. This is how it originally was, but of course we now need to be supporting registrar token transfers for TLDs.

We do it as follows: for non-TLD names, registry ownership is passed around. For TLD names, registrar token is passed around. As a courtesy, whenever transferring onto end user (buy in buy now offering, auction finalization, reclaim back when taking our auction down...), the Offering contract obtains registry ownership too and passes it to the end user.

### CODE REVIEW

File: src/name_bazaar/server/dev.cljs [hunk 1]
```diff
@@ -73,6 +73,7 @@
                             :endpoints {:port 6200
                                         :middlewares [logging-middlewares]}
                             :web3 {:port 8549}
+                            :db {:opts {:memory true}}
```
Comment (Siegrift) [inline src/name_bazaar/server/dev.cljs:76]: What this does?

File: compile-solidity.sh [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # TODO find a way to specify node_modules as an additional source root for solc
 # TODO or migrate to truffle and the entire compilation workflow will be changed
 ENS=$(readlink -f ./node_modules/@ensdomains)
-OZS=$(readlink -f ./node_modules/@ensdomains/ethregistrar/node_modules/openzeppelin-solidity)
```
Comment (Siegrift) [inline compile-solidity.sh:6]: Mhh, we did this in last PR, why revert it back? (I guess it doesn't change much though, so I am fine with either)

File: resources/public/contracts/src/Offering.sol [hunk 3]
```diff
@@ -176,11 +176,13 @@ contract Offering {
...
         private
     {
-        ens.setOwner(offering.node, _newOwner);
         if (isNodeTLDOfRegistrar()) {
             uint256 tokenId = uint256(offering.labelHash);
             BaseRegistrar registrar = BaseRegistrar(ens.owner(rootNode));
+            registrar.reclaim(tokenId, _newOwner);
             registrar.transferFrom(registrar.ownerOf(tokenId), _newOwner, tokenId);
+        } else {
+            ens.setOwner(offering.node, _newOwner);
```
Comment (Siegrift) [inline resources/public/contracts/src/Offering.sol:185]: Mhh, I remember this was in the else branch before, but we changed it in some PR (I think I even asked about this). We are reverting it back to `else` branch now. Why?

File: resources/public/contracts/src/Offering.sol [hunk 4]
```diff
@@ -209,7 +211,7 @@ contract Offering {
         if (isNodeTLDOfRegistrar()) {
             uint256 tokenId = uint256(offering.labelHash);
             address registrationOwner = BaseRegistrar(ens.owner(rootNode)).ownerOf(tokenId);
-            return ens.owner(offering.node) == address(this) && registrationOwner == address(this);
+            return registrationOwner == address(this);
```
Comment (Siegrift) [inline resources/public/contracts/src/Offering.sol:214]: Mhh, we only check who is the owner of registration now?

File: src/name_bazaar/server/contracts_api/registrar.cljs [hunk 5]
```diff
@@ -36,3 +36,5 @@
 (defn ens []
   (contract-call :name-bazaar-registrar :ens))
 
+(defn on-transfer [& args]
+  (apply contract-call :name-bazaar-registrar :Transfer args))
```
Comment (Siegrift) [inline src/name_bazaar/server/contracts_api/registrar.cljs:43]: do we need uppercase `T`? I guess not.

File: src/name_bazaar/ui/events/offerings_events.cljs [hunk 6]
```diff
@@ -652,16 +652,17 @@
...
-      [[:ens/set-owner {:ens.record/name name
-                        :ens.record/owner owner}]
-       (if (top-level-name? name)
-         [:name-bazaar-registrar/transfer {:ens.record/label (name-label name)
-                                           :ens.record/owner owner}
-          {:result-href (path-for :route.offerings/detail {:offering/address owner})
-           :on-tx-receipt-n [[:offerings.ownership/load [owner]]
-                             [:district0x.snackbar/show-message
-                              (gstring/format "Ownership of %s was transferred" name)]]}])]}))
+    {:dispatch
```
Comment (Siegrift) [inline src/name_bazaar/ui/events/offerings_events.cljs:655]: :+1:

File: src/name_bazaar/server/syncer.cljs [hunk 7]
```diff
@@ -26,11 +26,9 @@
 
 (defn node-owner? [offering-address {:keys [:offering/node :offering/top-level-name? :offering/label]}]
   (let [ens-owner (ens/owner {:ens.record/node node})]
-    (and (= ens-owner offering-address)
-         (if top-level-name?
-           (= (registrar/registration-owner {:ens.record/label label})
-              offering-address)
-           true))))
+    (if top-level-name?
```
Comment (Siegrift) [inline src/name_bazaar/server/syncer.cljs:29]: This is related to the change in smart contract. For, TLD we now only require the account to own the registration. Why this change?

File: src/name_bazaar/ui/subs/offerings_subs.cljs [hunk 8]
```diff
@@ -142,15 +142,13 @@
...
-               (if top-level-name?
-                 (:name-bazaar-registrar.registration/owner registrar-registration)
-                 true))
-      (= offering-address
-         (:ens.record/owner ens-record)
-         (if top-level-name?
-           (:name-bazaar-registrar.registration/owner registrar-registration)
-           offering-address)))))
+    (when
+      (if top-level-name?
```
Comment (Siegrift) [inline src/name_bazaar/ui/subs/offerings_subs.cljs]: Can we just directly compare if the owner of the domain (or registration) is equal to the offering address?

**Review by Siegrift — COMMENTED** [review]
Also, please rebase to current master.

I forced push it such that we are even with district master. In case you have problems, you can cherry pick the commits.

File: src/name_bazaar/server/dev.cljs (hunk 1 above)
Comment (ladislavdubravsky) [reply src/name_bazaar/server/dev.cljs:76]: It shows you a bit about how to configure the dev server sqlite database. If you choose `:memory false` a `db.db` file gets created in current folder. I found it the easiest way to inspect server db in debugging, since you can't have two connections to an in memory sqlite database.

File: compile-solidity.sh (hunk 2 above)
Comment (ladislavdubravsky) [reply compile-solidity.sh:6]: Originally we had a copypasted openzeppelin contract among our contracts. Then I replaced it with a npm dependency, which got installed into node_modules root. Ethregistrar however also depends on openzeppelin contracts and since it didn't like the one I had we got another openzeppelin in buried in `ethregistrar/node_modules`. These nuances were however lost on our compilation workflow, which only uses one location for OZS. When I noticed, I directed it to the buried OZS folder, since that's the one important for ethregistrar (NMBZ contracts are fine with either).

That was all good until I cleaned and reinstalled node_modules. On a clean install, since we no longer directly require any version of OZS, the OZS required by ethregistrar no longer gets installed into the buried ethregistrar folder, but to node_modules root instead. So that's the correct location, the old one wouldn't work on a clean npm install.

File: resources/public/contracts/src/Offering.sol (hunk 3 above)
Comment (ladislavdubravsky) [reply resources/public/contracts/src/Offering.sol:185]: Because the `registrar.reclaim` function has the nifty option to set the ENS owner to whomever you want if you're the token owner (i.e. reclaim to not just yourself). We're using it as you can see, in other words the other branch's `setOwner` is the `reclaim` call.

File: resources/public/contracts/src/Offering.sol (hunk 4 above)
Comment (ladislavdubravsky) [reply resources/public/contracts/src/Offering.sol:214]: Correct. The registration owner can reclaim ENS ownership at any time.

File: src/name_bazaar/server/syncer.cljs (hunk 7 above)
Comment (ladislavdubravsky) [reply src/name_bazaar/server/syncer.cljs:29]: The registration transfer is a must, since registration owner can reclaim ENS ownership at any time. ENS ownership tranfer is a courtesy. Ceasing to require ENS ownership at inopportune times enables us to transfer ownership into offering in one transaction only, which is the desired result.

**Comment by Siegrift** [comment]
I also rebased your PR upon my tests and found out that you need to add `[openzeppelin-solidity "2.3.0"]` back to project.
Otherwise the FE tests pass :)

File: resources/public/contracts/src/Offering.sol (hunk 3 above)
Comment (Siegrift) [reply resources/public/contracts/src/Offering.sol:185]: Oh, I see now. Thx

File: compile-solidity.sh (hunk 2 above)
Comment (Siegrift) [reply compile-solidity.sh:6]: I am not sure this works like you described. I did `lein npm i` and without the open zepelin dependency some of the functionality stops working.

File: src/name_bazaar/ui/subs/offerings_subs.cljs (hunk 8 above)
Comment (ladislavdubravsky) [reply src/name_bazaar/ui/subs/offerings_subs.cljs]: As it is the thing handles falsey values, returning nil. Note that it depends on subscriptions and is itself called in `subscribe`s, I think the nils represent missing / not yet present values and are desired in the [subscription workflows](https://purelyfunctional.tv/guide/re-frame-building-blocks/#subscriptions).

File: src/name_bazaar/server/contracts_api/registrar.cljs (hunk 5 above)
Comment (ladislavdubravsky) [reply src/name_bazaar/server/contracts_api/registrar.cljs:43]: Possibly not but I mirrored other contracts where they kept event names original and only function calls clojure-cased.

File: compile-solidity.sh (hunk 2 above)
Comment (ladislavdubravsky) [reply compile-solidity.sh:6]: Please check with `rm -rf node_modules` and `lein npm install`.

> I also rebased your PR upon my tests and found out that you need to add [openzeppelin-solidity "2.3.0"] back to project.
Otherwise the FE tests pass :)

:+1: but OZS is specified in `node_modules/@ensdomains/ethregistrar/package.json` and shouldn't be anywhere else.

Comment (ladislavdubravsky) [reply compile-solidity.sh:6]: For benefit of future readers, delete of `package-lock.json` was needed. Leiningen deletes `package.json` itself but keeps the lock file for management.

File: src/name_bazaar/ui/subs/offerings_subs.cljs (hunk 8 above)
Comment (Siegrift) [reply src/name_bazaar/ui/subs/offerings_subs.cljs]: Yeah, I understood it, but it took me some time. I think we can make it more readable.

Can you rewrite that using `let`? You can make the `not nil` check explicit and compare with the variable created in the let statement:
```clojure
    (let [node-owner (if top-level-name?
                       (:name-bazaar-registrar.registration/owner registrar-registration)
                       (:ens.record/owner ens-record))]
      (and node-owner (= offering-address node-owner)))))
```

**Review by Siegrift — APPROVED** [review]
LGTM :)

File: src/name_bazaar/ui/subs/offerings_subs.cljs (hunk 8 above)
Comment (ladislavdubravsky) [reply src/name_bazaar/ui/subs/offerings_subs.cljs]: Right that's nicer.
